### PR TITLE
Mio buffer writer

### PR DESCRIPTION
**Pull request by ywkaras**
better this time

### CODE REVIEW

File: iocore/eventsystem/I_MIOBufferWriter.h [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+class MIOBufferWriter : public ts::BufferWriter
+{
+public:
+  MIOBufferWriter(MIOBuffer &miob) : _miob(miob) { }
+
+  MIOBufferWriter &
+  write(const void *data_, size_t length) override
+  {
+    const char *data = reinterpret_cast<const char *>(data_);
```
Comment (SolidWallOfCode) [inline iocore/eventsystem/I_MIOBufferWriter.h]: `static_cast` doesn't work?

File: iocore/eventsystem/I_MIOBufferWriter.h [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+      return 0;
+    }
+
+    return iobbPtr->write_avail();
+  }
+
+  // Write the first n characters that have been placed in the auxiliary buffer.  This call invalidates the auxiliary buffer.
+  // This function should not be called if no auxiliary buffer is available.
+  //
+  virtual MIOBufferWriter &
```
Comment (SolidWallOfCode) [inline iocore/eventsystem/I_MIOBufferWriter.h]: Is `virtual` correct here? Maybe `override`? I don't think we'll have a subclass of this class.

File: iocore/eventsystem/I_MIOBufferWriter.h (hunk 1 above)
Comment (ywkaras) [reply iocore/eventsystem/I_MIOBufferWriter.h]: Probably.  I generally like to use the scariest cast possible when converting from a void pointer.  But converting to char is not scary really, so static_cast would be fine.

File: iocore/eventsystem/I_MIOBufferWriter.h (hunk 2 above)
Comment (ywkaras) [reply iocore/eventsystem/I_MIOBufferWriter.h]: Sorry brainfart.

**Comment by ywkaras** [comment]
Does it look like I'm using the write interface to MIOBuffer correctly?

**Comment by SolidWallOfCode** [comment]
Looking at it, actually not. You should be using `start()` instead of `buf()`. The former returns the first byte of not yet written space, while the latter is the buffer itself (which may already have data in it). So more like
```
int64_t avail = _miob->block_write_avail();
if (0 == avail) { _miob->add_block(); avail = _miob->block_write_avail(); }
int64_t to_write = std::min(length, avail);
memcpy(_miob->start(), buff, to_write);
length -= to_write;
```

**Comment by SolidWallOfCode** [comment]
Heh. One of the reasons I wanted this is because generically writing data to an MIOBuffer isn't obvious. It should be easy to say "I have a buffer of length N, write it all to this MIOBuffer".

**Comment by ywkaras** [comment]
OK, you're saying I should _not_ be getting a pointer to the first writable buffer then.

**Comment by SolidWallOfCode** [comment]
Yes. Because the first writeable buffer may not be empty.

**Comment by ywkaras** [comment]
So you assigned the person you doesn't know the non-obvious interface?

**Comment by SolidWallOfCode** [comment]
Nobody knows it, really.

**Comment by SolidWallOfCode** [comment]
To clarify a bit, there are several pointers to the first writable buffer.
* buf -> start of the buffer.
* start -> start of empty (unwritten) space in the buffer.
* end -> one past the end of the buffer

**Comment by ywkaras** [comment]
@SolidWallOfCode do you mean the member functions of IOBufferBlock in I_IOBuffer.h ?  You seem to be contradicting the code comments.  They say start() is the beginning of the "Inuse" section, which I assume means the already-written data.  They say end() is the end of the used space.  I assumed this was "end" in the STL sense, one past the last byte, the place to start adding more data.  It seems like one could possibly use member functions of MIOBuffer, which you seemed to suggest a few messages ago.  I'm going to try looking for examples of use in the code, and will perhaps die of shock if I find consistent usage.

**Comment by ywkaras** [comment]
Looking at iocore/net/SSLNetVConnection.cc , it would appear that my code is using the interface properly.

**Comment by SolidWallOfCode** [comment]
Yes, looks like you're correct. Oh well.

**Pull request closed without merging**